membershiporganization/migrations/0008_rename_realend_to_openended.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.29 on 2021-02-02 19:54
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+
+def map_new_type(apps, schema_editor):
+
+    MembershipOrganizationRealEnd = apps.get_model("membershiporganization", "MembershipOrganizationRealEnd")
+    db_alias = schema_editor.connection.alias
+
+    real_ends = MembershipOrganizationRealEnd.objects.using(db_alias).all()
+
+    for end in real_ends:
```

Comment:
@tlongers This logic will be used to convert the existing Boolean or null value for membership organization real end to the controlled Y/N/E vocabulary for open-endedness. Does this look correct to you, or would you suggest changes?

Reply:
I can't engage at this engineering level I'm afraid - that's way beyond my skill level. I can help frame a test however.

Reply:
Ah, got it! For context, real end is the value that we previously got from `unit:membership_last_cited_date_end`, which has gone away. (I think it might have been Y / N in the spreadsheet, but we converted that to a True / False Boolean in the import.) I'm converting the real end field in the model to an open ended field, as we discussed in the Google doc. The field conversion includes a step to transform existing values to the controlled vocab. Here's how I've told it to map those old Boolean values to new values: 

| real end (Boolean) | open ended (string) |
| - | - |
| True | N |
| False | Y |
| null | N |

Is that right? I'm fuzzy on what "real end" represents, actually, if you wouldn't mind providing a bit of explanation? In my afternoon brain fog, I assumed it was the inverse of open ended (i.e., if the membership has a real end, then it isn't open ended) but I could be misinterpreting!

Reply:
Thanks for laying it out clearly and tolerating our fiddlyness!  I've taken a look at [our docs](https://help.securityforcemonitor.org/en/latest/units.html#unit-membership-end-date) for `unit:membership_last_cited_date_end`, which says the boolean is mapped to the following meanings:
```
    Y: indicates that the membership ended on that date.
    N: indicates that the date is the date of last citation for the membership.
```
The naming of field doesn't actually have a concept of "open" _per se_,  at least not in the way that `unit:related_unit_open` does. The reason for this is that "memberships" in the way we use it as a organizational concept are abnormal and transient groupings of units: they wouldn't really be open-ended in the same way that a regular command structure is. Hence the difference in the data model from _the dark times_ ... before we saw the light!

So, a value of `Y` in `unit:membership_last_cited_date_end` means that the membership relationship ended on the exact date contained in `unit:membership_last_cited_date`. In these cases, the source we used will tell us that the particular membership ended on a specific date.  A value of `N` indicates that the data in `unit:membership_last_cited_date` is simply the last date we found; so the source we used is simply the last that we found that refers to the membership, but it doesn't tell us the relationship ended. Am I right to assume that for `unit:membership_last_cited_date_end` a value of `Y` is mapped to `True` and `N` to `False`?

Taking this forward into _the current times_ where `unit:membership_last_cited_date_end` is merged with `unit:related_unit_open`, the mapping looks like this:

unit:membership_last_cited_date.end|unit:related_unit_open|Which means that ...
--|--|--
`Y`|`E`| The date in `unit:related_unit_last_cited_date` is exact date that the two units stopped being related
`N`|`N`| The date we have in `unit:related_unit_last_cited_date` is the last date we have found, and we don't assume the relationship between units has continued
`null`|`N`
`(no equivalent)`|`Y`| The date we have in `unit:reated_unit_last_cited_date` is the last we have found, but we _do_ assume the relationship continues

I think in your mapping you might have it the wrong way around, unless I've also now become confused :/

_However_ when we pass you data in new format, I've already merged `unit:membership_last_cited_date_end` into `unit:related_unit_open` and transformed its values to match.

That help? Or ..

Reply:
This is perfect, @tlongers, thank you! I'll be sure to bookmark those docs. You're right that I've got it backwards. I'll update my mapping function accordingly.